include/cuco/detail/bloom_filter_kernels.cuh

```diff
@@ -0,0 +1,108 @@
+       tid += gridDim.x * block_size) {
+    new (&slots[tid]) atomic_slot_type{0};
+  }
+}
+
+/**
+ * @brief Inserts all keys in the range `[first, last)`.
+ *
+ * @tparam block_size The size of the thread block
+ * @tparam InputIt Device accessible input iterator whose `value_type` is
+ * convertible to the filter's `key_type`
+ * @tparam View Type of device view allowing access of filter storage
+ * @tparam Hash Unary callable type
+ * @param first Beginning of the sequence of keys
+ * @param last End of the sequence of keys
+ * @param view Mutable device view used to access the filter's slot storage
+ * @param hash The unary function to apply to hash each key
+ */
+template <std::size_t block_size, typename InputIt, typename View, typename Hash>
+__global__ void insert(InputIt first, InputIt last, View view, Hash hash)
+{
+  std::size_t tid = block_size * blockIdx.x + threadIdx.x;
+  auto it         = first + tid;
+
+  while (it < last) {
+    typename View::key_type const key{*it};
+    view.insert(key, hash);
+    it += gridDim.x * block_size;
+  }
+}
+
+/**
+ * @brief Indicates whether the keys in the range `[first, last)` are contained
+ * in the filter.
+ *
+ * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` exists
+ * in the filter.
+ *
+ * @tparam block_size The size of the thread block
+ * @tparam InputIt Device accessible input iterator whose `value_type` is
+ * convertible to the filter's `key_type`
+ * @tparam OutputIt Device accessible output iterator whose `value_type` is
+ * convertible to `bool`.
+ * @tparam View Type of device view allowing access of filter storage
+ * @tparam Hash Unary callable type
+ * @param first Beginning of the sequence of keys
+ * @param last End of the sequence of keys
+ * @param output_begin Beginning of the sequence of booleans for the presence of each key
+ * @param view Mutable device view used to access the filter's slot storage
+ * @param hash The unary function to apply to hash each key
+ */
+template <std::size_t block_size, typename InputIt, typename OutputIt, typename View, typename Hash>
+__global__ void contains(InputIt first, InputIt last, OutputIt output_begin, View view, Hash hash)
+{
+  std::size_t tid = block_size * blockIdx.x + threadIdx.x;
+  auto it         = first + tid;
+  __shared__ bool write_buffer[block_size];
+
+  while ((first + tid) < last) {
+    /*
+     * The ld.relaxed.gpu instruction used in view.contains causes L1 to
+     * flush more frequently, causing increased sector stores from L2 to global memory.
+     * By writing results to shared memory and then synchronizing before writing back
+     * to global, we no longer rely on L1, preventing the increase in sector stores from
+     * L2 to global and improving performance.
+     */
+    typename View::key_type const key{*(first + tid)};
+    write_buffer[threadIdx.x] = view.contains(key, hash);
+    __syncthreads();
```

Comment:
Interesting optimization! As far as I understand, it's unsafe to synchronize like this. If input size isn't a multiple of a `block_size`, `__syncthreads()` won't be called by all threads of a thread block. Or [this  information](https://stackoverflow.com/questions/6666382/can-i-use-syncthreads-after-having-dropped-threads) is outdated and it's fine :smile:

Reply:
This originally came from the `static_map`. @sleeepyjack did you test if this optimization actually made a difference on the `bloom_filter` too? But yes, we shouldn't be calling `__syncthreads` in a situation where not all threads will call it.

Reply:
Right, calling `__synchthreads` here is theoretically unsafe. I noticed this issue when implementing multimap but weird enough, compute sanitizer didn't throw an error in this case.

Reply:
I'd make sure to test on both V100 and A100.

Reply:
['../baseline_v100.json', '../direct_write_v100.json']

## [0] Tesla V100-PCIE-32GB

|  KeyType  |  SlotType  |  Operation  |  FilterScope  |  DataScope  |  NumInputs  |   NumBits    |  NumHashes  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|-----------|------------|-------------|---------------|-------------|-------------|--------------|-------------|------------|-------------|------------|-------------|------------|---------|----------|
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 461.451 ms |       0.02% | 461.523 ms |       0.02% |  72.063 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  46.441 ms |       0.52% |  46.450 ms |       0.53% |   9.029 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 480.755 ms |       0.01% | 480.731 ms |       0.01% | -23.465 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  48.724 ms |       0.78% |  48.721 ms |       0.77% |  -2.234 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 487.495 ms |       0.02% | 487.435 ms |       0.01% | -60.039 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  49.731 ms |       1.04% |  49.712 ms |       0.91% | -18.618 us |  -0.04% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 490.764 ms |       0.01% | 490.769 ms |       0.01% |   5.027 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  50.428 ms |       1.40% |  50.410 ms |       1.44% | -17.408 us |  -0.03% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 493.323 ms |       0.01% | 493.273 ms |       0.01% | -49.710 us |  -0.01% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  51.001 ms |       1.71% |  50.988 ms |       1.70% | -12.939 us |  -0.03% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 498.809 ms |       0.01% | 498.773 ms |       0.01% | -36.122 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  53.243 ms |       1.55% |  53.221 ms |       1.53% | -21.411 us |  -0.04% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 514.476 ms |       0.01% | 514.521 ms |       0.01% |  44.317 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  54.745 ms |       0.08% |  54.755 ms |       0.08% |   9.775 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 499.058 ms |       0.02% | 499.096 ms |       0.01% |  37.983 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  52.598 ms |       4.90% |  52.593 ms |       4.86% |  -5.492 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  1000000000  |      2      | 459.777 ms |       0.03% | 459.865 ms |       0.02% |  88.709 us |   0.02% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  1000000000  |      2      |  46.213 ms |       0.04% |  46.220 ms |       0.05% |   7.075 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  2000000000  |      2      | 479.117 ms |       0.01% | 479.118 ms |       0.01% |   1.196 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  2000000000  |      2      |  48.458 ms |       0.05% |  48.458 ms |       0.04% |  -0.372 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  3000000000  |      2      | 485.807 ms |       0.02% | 485.836 ms |       0.01% |  28.209 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  3000000000  |      2      |  49.420 ms |       0.04% |  49.410 ms |       0.05% | -10.333 us |  -0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  4000000000  |      2      | 489.105 ms |       0.01% | 489.222 ms |       0.01% | 117.015 us |   0.02% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  4000000000  |      2      |  50.045 ms |       0.03% |  50.063 ms |       0.04% |  18.246 us |   0.04% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  5000000000  |      2      | 491.695 ms |       0.01% | 491.736 ms |       0.01% |  40.863 us |   0.01% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  5000000000  |      2      |  50.597 ms |       0.02% |  50.606 ms |       0.05% |   8.472 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 10000000000  |      2      | 497.182 ms |       0.01% | 497.200 ms |       0.01% |  17.217 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 10000000000  |      2      |  52.637 ms |       0.15% |  52.655 ms |       0.17% |  17.594 us |   0.03% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 50000000000  |      2      | 513.013 ms |       0.01% | 513.048 ms |       0.01% |  34.735 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 50000000000  |      2      |  54.646 ms |       0.10% |  54.659 ms |       0.08% |  12.848 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 100000000000 |      2      | 541.184 ms |       0.01% | 541.213 ms |       0.02% |  29.413 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 100000000000 |      2      |  53.461 ms |       4.56% |  53.430 ms |       4.53% | -30.162 us |  -0.06% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 179.099 ms |       0.08% | 181.893 ms |       0.01% |   2.794 ms |   1.56% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  17.841 ms |       0.06% |  18.120 ms |       0.09% | 278.915 us |   1.56% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 185.438 ms |       0.03% | 188.114 ms |       0.02% |   2.676 ms |   1.44% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  18.490 ms |       0.05% |  18.741 ms |       0.06% | 251.167 us |   1.36% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 187.540 ms |       0.03% | 190.229 ms |       0.02% |   2.689 ms |   1.43% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  18.703 ms |       0.07% |  18.959 ms |       0.07% | 255.393 us |   1.37% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 188.541 ms |       0.02% | 191.286 ms |       0.03% |   2.745 ms |   1.46% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  18.809 ms |       0.06% |  19.058 ms |       0.08% | 249.173 us |   1.32% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 189.213 ms |       0.02% | 191.982 ms |       0.01% |   2.769 ms |   1.46% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  18.875 ms |       0.06% |  19.119 ms |       0.06% | 243.333 us |   1.29% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 190.541 ms |       0.02% | 193.308 ms |       0.03% |   2.767 ms |   1.45% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  19.188 ms |       5.12% |  19.446 ms |       5.03% | 258.563 us |   1.35% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 191.463 ms |       0.02% | 194.309 ms |       0.02% |   2.846 ms |   1.49% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  24.249 ms |       0.08% |  24.399 ms |       0.08% | 150.040 us |   0.62% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 247.120 ms |       0.01% | 249.836 ms |       0.01% |   2.716 ms |   1.10% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  24.519 ms |       0.03% |  24.750 ms |       0.03% | 231.034 us |   0.94% |   FAIL   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 461.439 ms |       0.02% | 461.492 ms |       0.02% |  52.967 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  47.668 ms |       4.71% |  47.656 ms |       4.65% | -12.101 us |  -0.03% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 480.727 ms |       0.02% | 480.706 ms |       0.01% | -21.135 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  49.840 ms |       4.12% |  49.831 ms |       4.14% |  -8.936 us |  -0.02% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 487.427 ms |       0.02% | 487.344 ms |       0.02% | -82.758 us |  -0.02% |   FAIL   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  50.813 ms |       3.65% |  50.797 ms |       3.66% | -15.639 us |  -0.03% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 490.753 ms |       0.01% | 490.727 ms |       0.01% | -26.528 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  51.493 ms |       3.53% |  51.494 ms |       3.52% |   0.837 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 493.263 ms |       0.01% | 493.274 ms |       0.01% |  10.795 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  52.036 ms |       3.18% |  52.038 ms |       3.19% |   2.421 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 498.721 ms |       0.01% | 498.713 ms |       0.02% |  -8.004 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  54.200 ms |       1.08% |  54.178 ms |       1.13% | -22.436 us |  -0.04% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 514.233 ms |       0.02% | 514.222 ms |       0.01% | -11.264 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  54.726 ms |       0.06% |  54.740 ms |       0.06% |  14.337 us |   0.03% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 527.539 ms |       1.80% | 527.513 ms |       1.80% | -26.622 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  49.918 ms |       0.04% |  49.910 ms |       0.04% |  -7.540 us |  -0.02% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  1000000000  |      2      | 459.804 ms |       0.02% | 459.843 ms |       0.02% |  38.910 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  1000000000  |      2      |  46.207 ms |       0.02% |  46.221 ms |       0.04% |  13.499 us |   0.03% |   FAIL   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  2000000000  |      2      | 479.157 ms |       0.01% | 479.100 ms |       0.01% | -56.505 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  2000000000  |      2      |  48.467 ms |       0.04% |  48.452 ms |       0.04% | -15.453 us |  -0.03% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  3000000000  |      2      | 485.875 ms |       0.02% | 485.811 ms |       0.01% | -64.878 us |  -0.01% |   FAIL   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  3000000000  |      2      |  49.438 ms |       0.15% |  49.438 ms |       0.13% |   0.371 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  4000000000  |      2      | 489.133 ms |       0.01% | 489.143 ms |       0.01% |  10.334 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  4000000000  |      2      |  50.103 ms |       0.34% |  50.108 ms |       0.35% |   4.747 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  5000000000  |      2      | 491.705 ms |       0.01% | 491.707 ms |       0.01% |   2.516 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  5000000000  |      2      |  50.678 ms |       0.53% |  50.676 ms |       0.51% |  -1.117 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 10000000000  |      2      | 497.234 ms |       0.02% | 497.191 ms |       0.01% | -42.910 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 10000000000  |      2      |  52.828 ms |       0.99% |  52.797 ms |       1.01% | -31.278 us |  -0.06% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 50000000000  |      2      | 514.068 ms |       0.66% | 514.019 ms |       0.65% | -49.161 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 50000000000  |      2      |  54.654 ms |       0.05% |  54.640 ms |       0.04% | -13.963 us |  -0.03% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 100000000000 |      2      | 541.215 ms |       0.02% | 541.174 ms |       0.01% | -40.871 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 100000000000 |      2      |  53.450 ms |       4.55% |  53.443 ms |       4.55% |  -6.889 us |  -0.01% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 180.668 ms |       0.08% | 183.493 ms |       0.02% |   2.825 ms |   1.56% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  18.329 ms |       6.60% |  18.615 ms |       6.39% | 285.708 us |   1.56% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 187.013 ms |       0.02% | 189.729 ms |       0.02% |   2.716 ms |   1.45% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  19.011 ms |       6.91% |  19.277 ms |       6.82% | 266.363 us |   1.40% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 189.116 ms |       0.03% | 191.815 ms |       0.03% |   2.698 ms |   1.43% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  19.251 ms |       7.10% |  19.494 ms |       6.88% | 243.161 us |   1.26% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 190.145 ms |       0.02% | 192.833 ms |       0.02% |   2.688 ms |   1.41% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  19.356 ms |       7.12% |  19.601 ms |       6.88% | 245.090 us |   1.27% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 190.846 ms |       0.03% | 193.543 ms |       0.01% |   2.697 ms |   1.41% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  19.425 ms |       7.10% |  19.671 ms |       6.86% | 246.233 us |   1.27% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 192.146 ms |       0.02% | 194.847 ms |       0.02% |   2.701 ms |   1.41% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  19.954 ms |       9.44% |  20.224 ms |       9.34% | 269.617 us |   1.35% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 193.089 ms |       0.02% | 195.797 ms |       0.03% |   2.708 ms |   1.40% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  24.440 ms |       0.06% |  24.598 ms |       0.08% | 158.135 us |   0.65% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 246.573 ms |       0.01% | 249.968 ms |       0.01% |   3.395 ms |   1.38% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  24.549 ms |       0.02% |  24.798 ms |       0.02% | 248.637 us |   1.01% |   FAIL   |

Reply:
['../baseline_a100.json', '../direct_write_a100.json']

## [0] NVIDIA A100 80GB PCIe

|  KeyType  |  SlotType  |  Operation  |  FilterScope  |  DataScope  |  NumInputs  |   NumBits    |  NumHashes  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|-----------|------------|-------------|---------------|-------------|-------------|--------------|-------------|------------|-------------|------------|-------------|------------|---------|----------|
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 154.549 ms |       0.37% | 154.551 ms |       0.36% |   1.953 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  15.578 ms |       0.99% |  15.577 ms |       1.02% |  -0.372 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 171.445 ms |       0.26% | 171.446 ms |       0.26% |   0.562 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  17.357 ms |       0.85% |  17.356 ms |       0.85% |  -0.247 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 177.142 ms |       0.24% | 177.137 ms |       0.24% |  -5.493 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  18.009 ms |       0.68% |  18.007 ms |       0.68% |  -1.353 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 180.017 ms |       0.24% | 180.027 ms |       0.23% |  10.799 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  18.380 ms |       0.47% |  18.378 ms |       0.48% |  -1.536 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 181.806 ms |       0.24% | 181.811 ms |       0.24% |   5.496 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  18.631 ms |       0.13% |  18.630 ms |       0.13% |  -0.228 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 185.644 ms |       0.29% | 185.638 ms |       0.29% |  -5.862 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  18.976 ms |       0.03% |  18.977 ms |       0.03% |   0.721 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 192.314 ms |       0.14% | 192.312 ms |       0.14% |  -1.953 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  19.248 ms |       0.03% |  19.252 ms |       0.02% |   4.293 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 192.727 ms |       0.01% | 192.720 ms |       0.01% |  -6.610 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  19.279 ms |       0.03% |  19.279 ms |       0.03% |   0.000 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  1000000000  |      2      | 155.908 ms |       1.19% | 155.888 ms |       1.19% | -20.015 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  1000000000  |      2      |  15.287 ms |       0.23% |  15.285 ms |       0.24% |  -1.272 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  2000000000  |      2      | 169.228 ms |       0.30% | 169.249 ms |       0.30% |  21.129 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  2000000000  |      2      |  17.061 ms |       0.25% |  17.057 ms |       0.25% |  -4.096 us |  -0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  3000000000  |      2      | 174.953 ms |       0.28% | 174.948 ms |       0.27% |  -4.748 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  3000000000  |      2      |  17.710 ms |       0.27% |  17.711 ms |       0.27% |   0.212 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  4000000000  |      2      | 177.843 ms |       0.27% | 177.854 ms |       0.28% |  10.241 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  4000000000  |      2      |  18.086 ms |       0.26% |  18.084 ms |       0.26% |  -1.572 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  5000000000  |      2      | 179.633 ms |       0.28% | 179.615 ms |       0.28% | -17.409 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  5000000000  |      2      |  18.346 ms |       0.17% |  18.344 ms |       0.17% |  -1.646 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 10000000000  |      2      | 183.491 ms |       0.33% | 183.493 ms |       0.33% |   2.050 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 10000000000  |      2      |  18.766 ms |       0.03% |  18.764 ms |       0.04% |  -1.934 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 50000000000  |      2      | 189.975 ms |       0.21% | 189.962 ms |       0.22% | -12.937 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 50000000000  |      2      |  19.040 ms |       0.04% |  19.038 ms |       0.04% |  -2.200 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 100000000000 |      2      | 190.613 ms |       0.01% | 190.583 ms |       0.02% | -29.606 us |  -0.02% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 100000000000 |      2      |  19.066 ms |       0.04% |  19.067 ms |       0.05% |   0.607 us |   0.00% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      |  55.813 ms |       4.36% |  55.941 ms |       4.40% | 128.465 us |   0.23% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |   5.309 ms |       2.11% |   5.321 ms |       2.11% |  11.362 us |   0.21% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      |  58.713 ms |       0.01% |  58.805 ms |       0.01% |  92.067 us |   0.16% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |   5.921 ms |       2.26% |   5.932 ms |       2.28% |  10.553 us |   0.18% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      |  60.641 ms |       0.01% |  60.738 ms |       0.01% |  97.652 us |   0.16% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |   6.125 ms |       2.52% |   6.134 ms |       2.52% |   9.328 us |   0.15% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      |  61.577 ms |       0.02% |  61.681 ms |       0.01% | 104.076 us |   0.17% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |   6.235 ms |       2.88% |   6.246 ms |       2.89% |  10.885 us |   0.17% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      |  62.136 ms |       0.01% |  62.247 ms |       0.02% | 111.244 us |   0.18% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |   6.313 ms |       3.27% |   6.323 ms |       3.25% |   9.318 us |   0.15% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      |  63.215 ms |       0.02% |  63.323 ms |       0.01% | 107.892 us |   0.17% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |   6.961 ms |       0.03% |   6.975 ms |       0.04% |  13.739 us |   0.20% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      |  65.379 ms |       3.83% |  65.480 ms |       3.85% | 101.003 us |   0.15% |   PASS   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |   7.058 ms |       0.04% |   7.072 ms |       0.05% |  13.413 us |   0.19% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      |  70.480 ms |       0.04% |  70.614 ms |       0.04% | 134.516 us |   0.19% |   FAIL   |
|    I32    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |   7.067 ms |       0.06% |   7.080 ms |       0.06% |  13.464 us |   0.19% |   FAIL   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 159.405 ms |       1.27% | 159.411 ms |       1.27% |   5.493 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  15.877 ms |       1.21% |  15.876 ms |       1.22% |  -0.576 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 173.715 ms |       0.86% | 173.715 ms |       0.86% |   0.282 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  17.645 ms |       0.91% |  17.645 ms |       0.91% |  -0.106 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 179.371 ms |       0.83% | 179.367 ms |       0.82% |  -4.095 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  18.290 ms |       0.56% |  18.291 ms |       0.57% |   0.658 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 182.221 ms |       0.81% | 182.226 ms |       0.81% |   4.372 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  18.628 ms |       0.04% |  18.627 ms |       0.03% |  -1.252 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 183.980 ms |       0.80% | 183.982 ms |       0.80% |   2.701 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  18.798 ms |       0.04% |  18.800 ms |       0.03% |   1.669 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 187.791 ms |       0.79% | 187.790 ms |       0.79% |  -1.583 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  19.140 ms |       0.04% |  19.138 ms |       0.03% |  -1.479 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 194.000 ms |       0.01% | 193.998 ms |       0.01% |  -1.770 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  19.412 ms |       0.03% |  19.410 ms |       0.03% |  -1.496 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 194.283 ms |       0.01% | 194.293 ms |       0.02% |  10.705 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  19.440 ms |       0.04% |  19.440 ms |       0.04% |   0.393 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  1000000000  |      2      | 154.812 ms |       1.51% | 154.811 ms |       1.51% |  -1.581 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  1000000000  |      2      |  15.308 ms |       0.63% |  15.309 ms |       0.62% |   1.179 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  2000000000  |      2      | 169.768 ms |       0.94% | 169.760 ms |       0.94% |  -7.819 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  2000000000  |      2      |  17.079 ms |       0.53% |  17.077 ms |       0.52% |  -1.946 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  3000000000  |      2      | 175.477 ms |       0.89% | 175.486 ms |       0.89% |   8.376 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  3000000000  |      2      |  17.732 ms |       0.44% |  17.734 ms |       0.45% |   1.624 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  4000000000  |      2      | 178.364 ms |       0.87% | 178.354 ms |       0.87% | -10.053 us |  -0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  4000000000  |      2      |  18.105 ms |       0.38% |  18.107 ms |       0.37% |   1.829 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  |  5000000000  |      2      | 180.147 ms |       0.86% | 180.142 ms |       0.86% |  -4.750 us |  -0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  |  5000000000  |      2      |  18.360 ms |       0.23% |  18.361 ms |       0.24% |   1.463 us |   0.01% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 10000000000  |      2      | 184.006 ms |       0.84% | 184.006 ms |       0.85% |   0.373 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 10000000000  |      2      |  18.766 ms |       0.03% |  18.769 ms |       0.04% |   2.844 us |   0.02% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 50000000000  |      2      | 190.272 ms |       0.01% | 190.295 ms |       0.02% |  23.084 us |   0.01% |   FAIL   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 50000000000  |      2      |  19.040 ms |       0.04% |  19.040 ms |       0.05% |   0.076 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  | 1000000000  | 100000000000 |      2      | 190.591 ms |       0.01% | 190.600 ms |       0.01% |   9.216 us |   0.00% |   PASS   |
|    I64    |    I32     |   INSERT    |     GMEM      |  REGISTERS  |  100000000  | 100000000000 |      2      |  19.071 ms |       0.03% |  19.070 ms |       0.03% |  -0.455 us |  -0.00% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      |  56.890 ms |       4.46% |  56.968 ms |       4.50% |  77.545 us |   0.14% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |   5.540 ms |       2.92% |   5.544 ms |       2.97% |   3.117 us |   0.06% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      |  60.497 ms |       0.52% |  60.551 ms |       0.58% |  53.527 us |   0.09% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |   6.129 ms |       3.23% |   6.135 ms |       3.30% |   6.144 us |   0.10% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      |  62.290 ms |       0.29% |  62.336 ms |       0.27% |  46.453 us |   0.07% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |   6.333 ms |       3.54% |   6.339 ms |       3.58% |   5.314 us |   0.08% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      |  63.188 ms |       0.29% |  63.248 ms |       0.38% |  60.043 us |   0.10% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |   6.446 ms |       3.84% |   6.452 ms |       3.87% |   5.960 us |   0.09% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      |  63.725 ms |       0.43% |  63.785 ms |       0.47% |  60.229 us |   0.09% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |   6.543 ms |       4.22% |   6.551 ms |       4.25% |   7.274 us |   0.11% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      |  64.977 ms |       1.39% |  65.047 ms |       1.47% |  69.538 us |   0.11% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |   7.114 ms |       0.04% |   7.124 ms |       0.04% |   9.836 us |   0.14% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      |  68.247 ms |       4.61% |  68.336 ms |       4.66% |  88.717 us |   0.13% |   PASS   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |   7.209 ms |       0.05% |   7.218 ms |       0.04% |   9.538 us |   0.13% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      |  71.961 ms |       0.03% |  72.067 ms |       0.03% | 106.030 us |   0.15% |   FAIL   |
|    I64    |    I32     |  CONTAINS   |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |   7.216 ms |       0.05% |   7.227 ms |       0.06% |  10.443 us |   0.14% |   FAIL   |